bio_hansel/subtyper.py

```diff
@@ -228,6 +228,24 @@ def parallel_query_reads(reads: List[Tuple[List[str], str]],
     outputs = [x.get() for x in res]
     return outputs
 
+def filter_by_kmer_fraction(df,min_kmer_frac=0.05):
```

Comment:
Hi @jrober84 I'm having trouble understanding what this function is supposed to do. 

Is kmer fraction supposed to be like the alternate allele fraction (AF) from variant calling?

What is a noisy kmer? Any kmer that is observed at a low frequency relative to the sum of frequencies of all kmers observed at that position? 

`kmer_freq / sum(kmer_freqs_at_position) < min_kmer_frac`

Reply:
Hey @peterk87, I have updated the description in the code to be a bit more clear. But essentially, the function determines the total number of positive and negative kmers covering a position and then determines the percentage of the total coverage each k-mer for the position contributes. In the case where only a single k-mer is present it will always be one and will not ever be filtered. But when both positive and negative are present it will see if the percentage contribution of each k-mer to the total is above the set threshold. If it isn't that specific k-mer gets filtered from the data frame and so the QA/QC module will never see it. I would say this process has some similarity to the AF for variant calling but also will allow the user to configure an acceptable "contamination" level for their sample which for their application is not an issue.

Reply:
Hi @jrober84 

I'm not sure this code is doing what your description says it's doing. I don't see any accessing of kmer frequency values. If I'm interpreting your description above and in the docstring correctly, you need to be calculating the sum of kmer frequencies at each position.  With `.value_counts()` on `refposition` you're simply getting a count of how many times you observe each `refposition` value (similar to Python's [Counter](https://docs.python.org/3/library/collections.html#counter-objects)). It also doesn't make sense that the `refposition` count is being divided by the `refposition` value. 

Also, when possible, I think it's a better idea to include all results for the detailed results report for debugging and troubleshooting rather than filtering those results out. I would instead just modify this line:

https://github.com/phac-nml/biohansel/blob/d20a00be813e72a678c3e8448ca783268b71c3cd/bio_hansel/subtyper.py#L289

Where instead of just getting the subset of results where `is_kmer_okay`, kmers that pass the kmer fraction threshold could also be filtered for:

```python
df[(df.is_kmer_freq_okay | (df.kmer_fraction >= subtyping_params.min_kmer_fraction))]
```

This way no results are removed from the detailed report, and only the "good" kmers are used to determine the subtype result.

So I would propose adding columns like `kmer_fraction` and maybe `total_refposition_kmer_frequency` and `does_pass_kmer_fraction_threshold`. This would be useful information for potential contamination detection.

Reply:
That's a good suggestion, I have updated the code to no longer be filtering the df for failed k-mers. This way all detected k-mers will appear in the detailed report but the QA/QC will be only on the pass list. Pass kmers must pass both the k-mer freq filter and k-mer frac filter. I have added the additional fields that you suggested for troubleshooting purposes.